tests/garage/envs/test_multi_task_env.py

```diff
@@ -0,0 +1,95 @@
+
+    def test_with_tasks_from_different_envs(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        mt_env = self._init_multi_task_env(envs)
+        assert mt_env.num_tasks == 2
+
+    def test_raise_exception_when_different_obs_space(self):
+        envs = ['CartPole-v0', 'Blackjack-v0']
+        with pytest.raises(ValueError):
+            _ = self._init_multi_task_env(envs)
+
+    def test_raise_exception_when_different_action_space(self):
+        envs = ['LunarLander-v2', 'LunarLanderContinuous-v2']
+        with pytest.raises(ValueError):
+            _ = self._init_multi_task_env(envs)
+
+    def test_default_active_task(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        mt_env = self._init_multi_task_env(envs)
+        assert mt_env.active_task == 0
+
+    def test_obs_space_and_action_space(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        mt_env = self._init_multi_task_env(envs)
+        task1 = TfEnv(env_name='CartPole-v0')
+        assert mt_env.observation_space.shape == task1.observation_space.shape
+        assert mt_env.action_space.shape == task1.action_space.shape
+
+    def test_sample_all_tasks(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        mt_env = self._init_multi_task_env(envs, sample_strategy='all')
+        tasks = mt_env.sample_tasks()
+        assert len(tasks) == 2
+        assert tasks[0] == 0 and tasks[1] == 1
+
+    def test_sample_random_tasks(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        mt_env = self._init_multi_task_env(envs, sample_strategy='random')
+        tasks = mt_env.sample_tasks(n_tasks=1)
+        assert len(tasks) == 1
+
+    def test_raise_exception_when_invalid_sample_strategy(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        with pytest.raises(ValueError):
+            _ = self._init_multi_task_env(envs, sample_strategy='not_random')
+
+    def test_set_task(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        mt_env = self._init_multi_task_env(envs)
+        mt_env.set_task(1)
+        assert mt_env.active_task == 1
+
+        mt_env.set_task(123)
+        assert mt_env.active_task == 123 % mt_env.num_tasks
+
+    def test_task_space(self):
+        envs = ['CartPole-v0', 'CartPole-v1']
+        mt_env = self._init_multi_task_env(envs)
+        bounds = mt_env.task_space.bounds
+        lb = np.zeros(2)
+        ub = np.ones(2)
+        assert (bounds[0] == lb).all() and (bounds[1] == ub).all()
+
+    def _init_multi_task_env(self, envs, sample_strategy='all'):
+        tasks = {}
+        for i, env in enumerate(envs):
+            task_name = 'task{0}'.format(i)
+            tasks[task_name] = {
+                'env': TfEnv,
+                'kwargs': {
+                    'env_name': env
+                },
+                'args': []
+            }
+
+        task_env = {k: v['env'] for k, v in tasks.items()}
+        task_kwargs = {k: v['kwargs'] for k, v in tasks.items()}
```

Comment:
this is my point about the args/kwargs interface being awkward, a stanza like this will appear in every launcher which uses this wrapper and it's ugly and hard to read.

we can instead make the interface "give me a list of envs" and perhaps provide some lightweight helpers for constructing those envs. we also might want a helper which allows you to define lazy-evaluated environment stubs (so we don't have to construct a bunch of environments in the local process, just so we can pickle them and re-construct them in a remote process).

Reply:
I agree.